### DISCUSSION TITLE
Can't load ResourceDictionary from source

### DISCUSSION THREAD

**Discussion by rickx** [AvaloniaUI/Avalonia#17549] [Q&A]
### Describe the bug

Hi,

I tried all possible variations and alternatives, but in the end, I think this is simply a bug. Loading Avalonia ResourceDictionary files from source files fails. And actually, I don't think it matters whether it's ResourceDictionaries or any other xaml files - it never reaches the "reading" of the files anyway.  

The error is
`Avalonia.Markup.Xaml.XamlLoadException: 'No precompiled XAML found for avares://AvaResourceDictionary_Loading/Resources/Dictionary.en-EN.axaml (baseUri: ), make sure to specify x:Class and include your XAML file as AvaloniaResource'`

so I think there are 2 possible reasons: 
* ResourceDicitionaries are only precompiled if found within other axaml files (windows/usercontrols) using the according tags.
* the csproj is not setting those files up correctly for being precompiled - setting them as AvaloniaResource seems to have as good as no effect (in this context).

There is also an _AvaloniaRuntimeXamlLoader_ but it's not accessible in 11.2.0 - maybe that would be the way of reading not-compiled xaml files(?)

### To Reproduce

Here's a small test project:

```public MainWindow()
{
    InitializeComponent();
    LoadResourceDictionary();
}
private void LoadResourceDictionary() { 
    var uri = new Uri("avares://AvaResourceDictionary_Loading/Resources/Dictionary.en-EN.axaml"); 
    var resourceDictionary = (ResourceDictionary)AvaloniaXamlLoader.Load(uri);
    Application.Current?.Resources.MergedDictionaries.Add(resourceDictionary); 
}
```

the resource file(s) may be empty - does not matter. Set to AvaloniaResource.

### Expected behavior

I should be able to load ResourceDictionaries from source and merge them at will. I used such a technique in WPF to conditionally load language and style dictionaries. 

### Avalonia version

11.2.0

### OS

Windows

### Additional context

Also tried ResourceInclude with no joy. It uses the same AvaloniaXamlLoader.Load in the end.

**Comment by rabbitism** [comment]
AvaloniaRuntimeXamlLoader is in `Avalonia.Markup.Xaml.Loader` nuget package. but it only loads raw xaml file, not assets.

**Comment by maxkatz6** [comment]
`ResourceInclude` is a typical API to load XAML dynamically.  `AvaloniaXamlLoader` would also work, just lower level.
Your problem is likely with trimming. Dynamic XAML loading can't work, if you app is trimmed (which it is by default since .NET 8+ I think).

If it's possible, you should always pre-include resources/styles in XAML, even as a keyed resource (i.e. not a merged resource includes, but just `<ResourceInclude x:Key="PreincludedResources" Source="Resources.xaml" />` a resource entry, which then can be read in runtime).

**Comment by timunie** [comment]
PreincludedResources

Huh? Didn't know this is a thing. If it's not yet in the docs, it should be mentioned there.

**Comment by maxkatz6** [comment]
Just like any other types, you can include any object in resources :)
Including ResourceInclude.

**Comment by rickx** [comment]
could you elaborate on accessing them at runtime? 
So, I added them as keyed resources. And that btw works only from App.axaml, not other controls. 
Now at runtime I find them all nicely loaded under app.Resources. Which, for my goal, means I have way more than I need. So I would pick the needed ones from there, merge them with app.Resources.MergedDictionaries.Add(,,,) and at that point, I would remove the unneeded Resources. Is that right? Are there other/better ways?

**Comment by maxkatz6** [comment]
@rickx just like with any other keyed resource, it doesn't have to be defined in App.xaml. It can be inside any other XAML control, or nested merged dictionary.
But you are right about how to use it, `MergedDictionaries.Add` and `MergedDictionaries.Remove` in runtime, using ResourceInclude read from the resources themselves.

**Comment by maxkatz6** [comment]
As an example, we use similar technique for Compact styles of FluentTheme:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/FluentTheme.xaml#L21-L22

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/FluentTheme.xaml.cs#L32
